File: sample/src/main/swift/Sources/SampleAppCore/UInt8Test.swift
//
// Created by Andrew on 2/4/18.
//

import Foundation

public enum UInt8Enum: UInt8, Codable {
    case one
    case two
    case three
}

public struct UInt8OptionsSet: OptionSet, Codable {

    public let rawValue: UInt8

    public init(rawValue: UInt8) {
        self.rawValue = rawValue
    }

    static let one = UInt8OptionsSet(rawValue: 1 << 0)
    static let two = UInt8OptionsSet(rawValue: 1 << 1)
    static let three = UInt8OptionsSet(rawValue: 1 << 2)
}

public struct UInt8TestStruct: Codable, Hashable {
    public var zero: UInt8 = UInt8.zero
    public var max: UInt8 = UInt8.max
    public var min: UInt8 = UInt8.min
    public var optional: UInt8? = UInt8.zero
    public var optionalNil: UInt8? = nil
}

public protocol UInt8TestParamProtocol {
    func testParam(_ param: UInt8) -> Bool
}

public protocol UInt8TestReturnTypeProtocol {
    func testReturnType() -> UInt8
}

public protocol UInt8TestOptionalParamProtocol {
    func testOptionalParam(_ param: UInt8?) -> Bool
}

public protocol UInt8TestOptionalReturnTypeProtocol {
    func testOptionalReturnType() -> UInt8?
}

public typealias UInt8Block = (_ value: UInt8) -> UInt8
public typealias OptionalUInt8Block = (_ value: UInt8?) -> UInt8?

public class UInt8Test {

    public static func testZero() -> UInt8 {
        return 0
    }

    public static func testMin() -> UInt8 {
        return UInt8.min
    }

    public static func testMax() -> UInt8 {
        return UInt8.max
    }

    public static func testParam(_ param: UInt8) -> Bool {
        return param == UInt8.max
    }

    public static func testReturnType() -> UInt8 {
        return UInt8.max
    }

    public static func testOptionalParam(_ param: UInt8?) -> Bool {
        return param == UInt8.max
    }

    public static func testOptionalReturnType() -> UInt8? {
        return UInt8.max
    }

    public static func testProtocolParam(_ callback: UInt8TestParamProtocol) -> Bool {
        return callback.testParam(UInt8.max)
    }

    public static func testProtocolReturnType(_ callback: UInt8TestReturnTypeProtocol) -> UInt8 {
        return callback.testReturnType()
    }

    public static func testProtocolOptionalParam(_ callback: UInt8TestOptionalParamProtocol) -> Bool {
        return callback.testOptionalParam(UInt8.max)
    }

    public static func testProtocolOptionalReturnType(_ callback: UInt8TestOptionalReturnTypeProtocol) -> UInt8? {
        return callback.testOptionalReturnType()
    }

    public static func testEncode() ->  UInt8TestStruct {
        return UInt8TestStruct()
    }

    public static func testDecode(_ value: UInt8TestStruct) -> Bool {
        return value == UInt8TestStruct()
    }

    public static func testEnumEncode(_ rawValue: UInt8) -> UInt8Enum {
        switch rawValue {
        case UInt8Enum.one.rawValue: return UInt8Enum.one
        case UInt8Enum.two.rawValue: return UInt8Enum.two
        case UInt8Enum.three.rawValue: return UInt8Enum.three
        default: fatalError("Can't find enum with rawValue \(rawValue)")
        }
    }

    public static func testEnumDecode(_ enumValue: UInt8Enum) -> UInt8 {
        return enumValue.rawValue
    }

    public static func testOptionSetEncode(_ rawValue: UInt8) -> UInt8OptionsSet {
        return UInt8OptionsSet(rawValue: rawValue)
    }

    public static func testOptionSetDecode(_ optionSet: UInt8OptionsSet) -> UInt8 {
        return optionSet.rawValue
    }

    public static func testBlock(_ block: UInt8Block) -> Bool {
        let value = block(0)
        return value == 0
    }

    public static func testOptionalBlock(_ block: OptionalUInt8Block) -> Bool {
        let value = block(nil)
        return value == nil
    }

}

File: sample/src/main/java/com/readdle/swiftjava/sample/UInt8Test.kt
package com.readdle.swiftjava.sample

import com.readdle.codegen.anotation.*
import java.lang.annotation.Native

/*
 *  Byte representation in Int8 and UInt8
 *  Int8:    | 0 | 0 | 0 | 0 | 0 | 0 | 0 | Sign bit: 0 | = 0
 *  UInt8:   | 0 | 0 | 0 | 0 | 0 | 0 | 0 |           0 | = 0
 *  Int8:    | 1 | 1 | 1 | 1 | 1 | 1 | 1 | Sign bit: 0 | = 127
 *  UInt8:   | 1 | 1 | 1 | 1 | 1 | 1 | 1 |           0 | = 127
 *  Int8:    | 1 | 1 | 1 | 1 | 1 | 1 | 1 | Sign bit: 1 | = -1
 *  UInt8:   | 1 | 1 | 1 | 1 | 1 | 1 | 1 |           1 | = 255
 */
private const val MIN_VALUE: Byte = 0   // | 0 | 0 | 0 | 0 | 0 | 0 | 0 | Sign bit: 0 |
private const val MAX_VALUE: Byte = -1  // | 1 | 1 | 1 | 1 | 1 | 1 | 1 | Sign bit: 1 |

@SwiftValue
enum class UInt8Enum(@Unsigned val rawValue: Byte) {

    ONE(0), TWO(1), THREE(2);

    companion object {

        private val values = HashMap<Byte, UInt8Enum>()

        @JvmStatic
        fun valueOf(rawValue: Byte): UInt8Enum {
            return values[rawValue]!!
        }

        init {
            enumValues<UInt8Enum>().forEach {
                values[it.rawValue] = it
            }
        }
    }

}

@SwiftValue
data class UInt8OptionsSet(@Unsigned val rawValue: Byte = 0) {
    companion object {
        @JvmStatic
        val  one = UInt8OptionsSet(1)
        @JvmStatic
        val  two = UInt8OptionsSet(2)
        @JvmStatic
        val  three = UInt8OptionsSet(4)
    }
}

@SwiftValue
data class UInt8TestStruct(@Unsigned var zero: Byte = 0,
                           @Unsigned var max: Byte = MAX_VALUE,
                           @Unsigned var min: Byte = MIN_VALUE,
                           @Unsigned var optional: Byte? = 0,
                           @Unsigned var optionalNil: Byte? = null)

@SwiftReference
class UInt8Test private constructor() {

    @SwiftDelegate(protocols = ["UInt8TestParamProtocol"])
    interface UInt8ParamProtocol {
        @SwiftCallbackFunc
        fun testParam(@Unsigned param: Byte): Boolean
    }

    @SwiftDelegate(protocols = ["UInt8TestReturnTypeProtocol"])
    interface UInt8ReturnTypeProtocol {
        @SwiftCallbackFunc @Unsigned
        fun testReturnType(): Byte
    }

    @SwiftDelegate(protocols = ["UInt8TestOptionalParamProtocol"])
    interface UInt8OptionalParamProtocol {
        @SwiftCallbackFunc
        fun testOptionalParam(@Unsigned param: Byte?): Boolean
    }

    @SwiftDelegate(protocols = ["UInt8TestOptionalReturnTypeProtocol"])
    interface UInt8OptionalReturnTypeProtocol {
        @SwiftCallbackFunc @Unsigned
        fun testOptionalReturnType(): Byte?
    }

    @SwiftBlock("(UInt8) -> UInt8")
    interface UInt8Block {
        @Unsigned
        fun call(@Unsigned value: Byte): Byte
    }

    @SwiftBlock("(UInt8?) -> UInt8?")
    interface OptionalUInt8Block {
        @Unsigned
        fun call(@Unsigned value: Byte?): Byte?
    }

    companion object {
        @JvmStatic @Unsigned
        external fun testZero(): Byte

        @JvmStatic @Unsigned
        external fun testMin(): Byte

        @JvmStatic @Unsigned
        external fun testMax(): Byte

        @JvmStatic
        external fun testParam(@Unsigned param: Byte): Boolean

        @JvmStatic @Unsigned
        external fun testReturnType(): Byte

        @JvmStatic
        external fun testOptionalParam(@Unsigned param: Byte?): Boolean

        @JvmStatic @Unsigned
        external fun testOptionalReturnType(): Byte?

        @JvmStatic
        external fun testProtocolParam(callback: UInt8ParamProtocol): Boolean

        @JvmStatic
        external fun testProtocolReturnType(callback: UInt8ReturnTypeProtocol): Byte

        @JvmStatic
        external fun testProtocolOptionalParam(callback: UInt8OptionalParamProtocol): Boolean

        @JvmStatic
        external fun testProtocolOptionalReturnType(callback: UInt8OptionalReturnTypeProtocol): Byte?

        @JvmStatic
        external fun testEncode():  UInt8TestStruct

        @JvmStatic
        external fun testDecode(value: UInt8TestStruct): Boolean

        @JvmStatic
        external fun testEnumEncode(@Unsigned rawValue: Byte) : UInt8Enum

        @JvmStatic @Unsigned
        external fun testEnumDecode(enum: UInt8Enum) : Byte

        @JvmStatic
        external fun testOptionSetEncode(@Unsigned rawValue: Byte) : UInt8OptionsSet

        @JvmStatic @Unsigned
        external fun testOptionSetDecode(enum: UInt8OptionsSet) : Byte
        
        @JvmStatic
        external fun testBlock(@SwiftBlock block: UInt8Block): Boolean

        @JvmStatic
        external fun testOptionalBlock(@SwiftBlock block: OptionalUInt8Block): Boolean
    }

    @Native
    var nativePointer: Long = 0

    external fun release()

}